### PR TITLE
l4 ace node-utils

### PR DESCRIPTION
**Pull request by saqibraza**

### CODE REVIEW

**Comment by bansalpradeep** [comment]
test comment

**Comment by saqibraza** [comment]
got it!

**Comment by mikewiebe** [comment]
@saqibraza Quite a few rubocop errors here, please fix.

File: lib/cisco_node_utils/ace.rb [hunk 1]
```diff
@@ -99,22 +108,53 @@ def ace_set(attrs)
...
         end
+        @get_args = @set_args
       end
-      set_args_keys(attrs)
       config_set('acl', cmd, @set_args)
     end
 
+    # UTILITY FUNCTIONS
+    # -----------------
+    def extract_value(prop, prefix)
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/ace.rb]: In most of these cases `prop` is the same as `prefix` so you could just make `prefix` optional to simplify the callers:

```
  def extract_value(prop, prefix=nil)
    prefix = prop if prefix.nil?
```

Come to think of it, `prefix` may not be the best choice of var names. Maybe `keyword` or `cli` instead?

File: lib/cisco_node_utils/ace.rb [hunk 2]
```diff
@@ -99,22 +108,53 @@ def ace_set(attrs)
...
+      match1 = ace_get
+      return nil if match1.nil?
+      return nil if match1.names.inclucde?(prop) == false
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/ace.rb]: `/inclucde/include/`

File: lib/cisco_node_utils/ace.rb [hunk 3]
```diff
@@ -99,22 +108,53 @@ def ace_set(attrs)
...
+      regexp = Regexp.new("#{Regexp.escape(prefix)} (?<extracted>*)")
+      match2 = regexp.match(match1[prop])
+      return nil if match2.nil?
+      match2[:extracted]
+    end
+
+    def attach_prefix(val, prop, prefix)
+      @set_args[prop] = val.nil? || val == '' ? val : prefix + ' ' + val
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/ace.rb]: A bit cleaner:

```
      @set_args[prop] = val.to_s.empty? ? val : "#{prefix} #{val}"
```

File: lib/cisco_node_utils/ace.rb [hunk 4]
```diff
@@ -192,5 +232,99 @@ def dst_port
...
+    end
+
+    def established
+      match = ace_get
+      return nil if match.nil?
+      match.names.include?('established') ? true : false
+    end
+
+    def established=(established)
+      @set_args[:established] = established == true ? 'established' : ''
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/ace.rb:264]:
```
      @set_args[:established] = established ? 'established' : ''

```

File: lib/cisco_node_utils/ace.rb [hunk 5]
```diff
@@ -99,22 +108,53 @@ def ace_set(attrs)
...
+    def extract_value(prop, prefix)
+      match1 = ace_get
+      return nil if match1.nil?
+      return nil if match1.names.inclucde?(prop) == false
+      regexp = Regexp.new("#{Regexp.escape(prefix)} (?<extracted>*)")
+      match2 = regexp.match(match1[prop])
+      return nil if match2.nil?
+      match2[:extracted]
+    end
+
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/ace.rb:162]: Some comments would be helpful on `extract_value` and `attach_prefix`

File: lib/cisco_node_utils/ace.rb [hunk 6]
```diff
@@ -192,5 +232,99 @@ def dst_port
...
+    def log
+      match = ace_get
+      return nil if match.nil?
+      match.names.include?('log') ? true : false
+    end
+
+    def log=(log)
+      @set_args[:log] = log == true ? 'log' : ''
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/ace.rb:334]: `@set_args[:log] = log ? 'log' : ''`

File: tests/test_ace.rb [hunk 7]
```diff
@@ -87,9 +87,115 @@ def test_create_destroy_ace_one
...
+                 attr_v4_flags_2, attr_v4_flags_3, attr_v4_flags_4,
+                 attr_v4_est_1, attr_v4_est_2, attr_v4_dscp_1, attr_v4_dscp_2,
+                 attr_v4_dscp_3, attr_v4_tcp_comb_1, attr_v4_tcp_comb_2],
+      'ipv6' => [attr_v6_1, attr_v6_2, attr_v6_3, attr_v6_flags_1,
+                 attr_v6_flags_2, attr_v6_flags_3, attr_v6_flags_4,
+                 attr_v6_est_1, attr_v6_est_2, attr_v6_dscp_1, attr_v6_dscp_2,
+                 attr_v6_dscp_3, attr_v6_tcp_comb_1, attr_v6_tcp_comb_2],
     }
 
     %w(ipv4 ipv6).each do |afi|
```
Comment (chrisvanheuveln) [inline tests/test_ace.rb:202]: This method is too long for rubocop - just wrap it in a disable/enable to make rubocop happy, ie:

```
  # TESTS
  # rubocop:disable Metrics/MethodLength
   def test_create_destroy_ace_one

       (snip)

   end
  # rubocop:enable Metrics/MethodLength
```

File: lib/cisco_node_utils/ace.rb (hunk 6 above)
Comment (saqibraza) [inline lib/cisco_node_utils/ace.rb:334]: The correction will set log when empty string is passed as argument, which we do not want.

File: lib/cisco_node_utils/ace.rb (hunk 4 above)
Comment (saqibraza) [inline lib/cisco_node_utils/ace.rb:264]: The correction will set established when empty string is passed as argument, which we do not want.

**Pull request merged**